### DISCUSSION TITLE
How to extract warmup samples?

### DISCUSSION THREAD

**Discussion by Madhav-Kanda** [blackjax-devs/blackjax#549] [Q&A] [answered]
Is it possible to extract warmup samples in Blackjax?

**Comment by Madhav-Kanda** [comment, marked as answer]
It was simple to get it. Mentioning it here in case someone is not able to figure it out:

Refer to the [example ](https://blackjax-devs.github.io/blackjax/examples/quickstart.html#use-stan-s-window-adaptation) provided on the blackjax site for window adaptation based on stan. To obtain the adaptation info (warmup samples) modify the code as follows:
```
warmup = blackjax.window_adaptation(blackjax.nuts, logdensity)
(state, parameters), warmup_samples = warmup.run(rng_key, initial_position, num_steps=1000)
```

In the above code, the variable named ```warmup_samples``` will contain the warmup samples.

**Comment by junpenglao** [comment]
🤖 *This reply was drafted with Claude and reviewed by me (@junpenglao).*

To also keep the per-step warmup states, use `return_all_adapt_info=True` (or a custom `adaptation_info_fn`) when constructing the adaptation:

```python
warmup = blackjax.window_adaptation(blackjax.nuts, logdensity_fn, adaptation_info_fn=blackjax.adaptation.base.return_all_adapt_info)
(state, params), info = warmup.run(rng_key, init_position, num_steps=1000)
# info now carries the adaptation trace
```

Note the current `run` returns `(state, parameters), info` — you build the kernel yourself with `blackjax.nuts(logdensity_fn, **params)`.
